src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -192,7 +195,24 @@ private void CaptureVariable(Symbol variable, SyntaxNode syntax)
             }
             else
             {
-                _variablesToHoist.Add(variable);
+                if (_variablesToHoist.Add(variable) && variable is LocalSymbol local && _boundRefLocalInitializers.TryGetValue(local, out var variableInitializer))
+                    CaptureRefInitializer(variableInitializer, syntax);
+            }
+        }
+
+        private void CaptureRefInitializer(BoundExpression variableInitializer, SyntaxNode syntax)
+        {
+            switch (variableInitializer)
+            {
+                case BoundLocal { LocalSymbol: var symbol }:
+                    CaptureVariable(symbol, syntax);
+                    break;
+                case BoundParameter { ParameterSymbol: var symbol }:
```

Comment:
Not one that would crash currently (without the fix), as I believe we always capture parameters that are used in any way.  That's because the state machine starts in a "cold" state, so hoisting the parameter is necessary for it to reach its use site.  So there is no way to construct an observable crash in the current compiler.

Though, hang on, I guess I could assign to the parameter before its use to repro the issue.  I will attempt to construct a case that crashes currently and exercises this to fix it.

---
In reply to: [318840878](https://github.com/dotnet/roslyn/pull/38196#discussion_r318840878) [](ancestors = 318840878)

Reply:
No, we hoist any parameter that is written to as well.  I will add a test in any case so if we stop hoisting parameters just because they are written then this won't be a problem.

---
In reply to: [319306652](https://github.com/dotnet/roslyn/pull/38196#discussion_r319306652) [](ancestors = 319306652,318840878)